byteps/server/server.cc

```diff
@@ -80,26 +84,61 @@ void BytePSServerEngineThread(int i) {
     CHECK(msg.dst);
     CHECK(msg.src);
 
+    auto iter = compressor_map_.find(msg.key);
+    if (iter != compressor_map_.end()) {
+      // compress
+      if (msg.ops == ALL_RECV) {
+        common::compressor::tensor_t grad(reinterpret_cast<char*>(msg.src),
+                                          msg.len, msg.type.dtype);
+        auto compressed = iter->second->Compress(grad);
+        // 1. compress
+        auto& updates = update_buf_[msg.key];
+        updates.merged.tensor = compressed.data;
+        updates.merged.len = compressed.size;
+      } else {  // decompress
+        auto compressed_len = msg.sarray.lens[0];
+        CHECK_LE(compressed_len, msg.len);
+        common::compressor::tensor_t compressed(
+            reinterpret_cast<char*>(msg.src), compressed_len, msg.type.dtype);
+        auto decompressed = iter->second->Decompress(compressed);
+        msg.src = decompressed.data;
+      }
+    } else {
+      if (msg.ops == ALL_RECV) {
+        // 2. no compress
+        auto& updates = update_buf_[msg.key];
+        updates.merged.tensor = reinterpret_cast<char*>(msg.src);
+        updates.merged.len = msg.len;
+      }
+    }
+
     bool is_debug = (debug_mode_ && (debug_key_ == msg.key));
     switch (msg.ops) {
-      case COPY_MERGED: {
+      case COPY_FIRST: {
         if (is_debug) {
           std::lock_guard<std::mutex> lock(debug_mu_);
           LOG(INFO) << "stage: ENGINE_COPY_MERGED_TO_STORE_BEFORE \t"
                     << "dst: " << DEBUG_PRINT_TENSOR_VALUE(msg.dst) << "\t"
                     << "src: " << DEBUG_PRINT_TENSOR_VALUE(msg.src) << "\t"
-                    << "dst_addr: " << DEBUG_PRINT_TENSOR_ADDRESS(msg.dst) << "\t"
-                    << "src_addr: " << DEBUG_PRINT_TENSOR_ADDRESS(msg.src) << "\t";
+                    << "dst_addr: " << DEBUG_PRINT_TENSOR_ADDRESS(msg.dst)
+                    << "\t"
+                    << "src_addr: " << DEBUG_PRINT_TENSOR_ADDRESS(msg.src)
+                    << "\t";
         }
         bps_reducer_->copy(msg.dst, msg.src, msg.len);
         if (is_debug) {
           std::lock_guard<std::mutex> lock(debug_mu_);
           LOG(INFO) << "stage: ENGINE_COPY_MERGED_TO_STORE_AFTER \t"
                     << "dst: " << DEBUG_PRINT_TENSOR_VALUE(msg.dst) << "\t"
                     << "src: " << DEBUG_PRINT_TENSOR_VALUE(msg.src) << "\t"
-                    << "dst_addr: " << DEBUG_PRINT_TENSOR_ADDRESS(msg.dst) << "\t"
-                    << "src_addr: " << DEBUG_PRINT_TENSOR_ADDRESS(msg.src) << "\t";
+                    << "dst_addr: " << DEBUG_PRINT_TENSOR_ADDRESS(msg.dst)
+                    << "\t"
+                    << "src_addr: " << DEBUG_PRINT_TENSOR_ADDRESS(msg.src)
+                    << "\t";
         }
+      } break;
+
+      case ALL_RECV: {
```

Comment:
`updates.merged.tensor` should be initialized under this case I think.

Reply:
It is already done in L107-L112.

Reply:
if compression is turned off L107-L112 won't be reached, right? I added `assert(1 == 2);` right after L88, it was never hit. Is that the right behavior?

Reply:
L107-L112 is reached when the compression is turned off. And when it is turned off, that assert should never be hit.

Reply:
Compression is tensor-wised. That means some gradients may be compressed but others not. To the extreme, all are compressed or none is compressed.

Reply:
> L107-L112 is reached when the compression is turned off

You mean turned on?

Reply:
> > L107-L112 is reached when the compression is turned off
> 
> You mean turned on?

```iter != compressor_map_.end()``` means the compressor is registered under the key. And L107-L112 is reached when there is no compressor registered, which means compression is turned off for that gradient of the key.